[processor/isolationforest] Fix goroutine leak in modelUpdateLoop shutdown

VenuEmmadi opened the pull request:
#### Description
This PR fixes goroutine leaks in the isolation forest processor caused by improper shutdown coordination of background `modelUpdateLoop` goroutines.

The processor was starting background goroutines for periodic model updates but not properly waiting for them to terminate during shutdown, causing `goleak` test failures. The tests would pass functionally but fail due to leaked goroutines remaining active after test completion.

**Root Cause:**
- `newIsolationForestProcessor` starts `go processor.modelUpdateLoop()` 
- `Shutdown()` method signals stop but doesn't wait for goroutine termination
- Tests don't call `Shutdown()` on created processors

**Fix:**
- Added `shutdownWG sync.WaitGroup` for proper goroutine tracking
- Modified `Shutdown()` to wait for all background goroutines to complete
- Updated tests to call `Shutdown()` during cleanup
- No functional changes to processor behavior

#### Link to tracking issue
Fixes #42675 

#### Testing
- All existing tests now pass without goroutine leaks
- `goleak` tests pass successfully  
- Added proper shutdown cleanup to all test functions that create processors
- Verified no regression in processor functionality
- Background goroutines properly terminate in both test and production environments

**Files tested:**
- `processor_test.go` - Fixed 5 test functions to call `Shutdown()`
- `factory_test.go` - Already had proper shutdown calls
- `integration_test.go` - Verified shutdown coordination works

#### Documentation
- Updated inline code comments explaining shutdown coordination
- No user-facing documentation changes needed (internal fix only)
- Proper goroutine lifecycle management maintains existing API

VenuEmmadi committed:
fix(processor/isolationforest): fix goroutine leak in modelUpdateLoop shutdown

- Add shutdownWG sync.WaitGroup for proper goroutine coordination
- Fix Shutdown() method to wait for background goroutines to complete
- Add proper cleanup in modelUpdateLoop with defer shutdownWG.Done()
- Fix tests to call Shutdown() to prevent goroutine leaks
- Resolves goleak test failures

Fixes goroutine leak detected by goleak in test runs.

VenuEmmadi committed:
Added file to chloggen

VenuEmmadi committed:
Fixed lint

VenuEmmadi committed:
fix(processor/isolationforest): fix goroutine leak in modelUpdateLoop

- Add proper shutdown coordination with sync.WaitGroup
- Update tests to use t.Cleanup() and t.Context()
- Resolves goleak test failures

axw commented on processor/isolationforestprocessor/processor_test.go:
```diff
@@ -96,6 +96,12 @@ func Test_newIsolationForestProcessor_Basic(t *testing.T) {
 
+	// Add shutdown cleanup
+	t.Cleanup(func() {
+		shutdownErr := p.Shutdown(t.Context())
```
```suggestion
		shutdownErr := p.Shutdown(context.WithoutCancel(t.Context()))
```

The t.Context method returns a context that is cancelled when the test ends, before cleanups are invoked. This has been the cause of flaky tests in other components recently.

I see that the Shutdown method ignores context (it shouldn't), but even so it would be good to make this change to avoid others copy & pasting the same thing and causing flaky tests in other components.

axw commented on processor/isolationforestprocessor/processor.go:
```diff
@@ -103,6 +104,10 @@ func newIsolationForestProcessor(config *Config, logger *zap.Logger) (*isolation
 
 	processor.updateTicker = time.NewTicker(updateFreq)
+
+	// Properly track the background goroutine with WaitGroup
+	processor.shutdownWG.Add(1)
+
 	go processor.modelUpdateLoop()
```
```suggestion
	// Properly track the background goroutine with WaitGroup
	processor.shutdownWG.Add(1)
	go func() {
		defer processor.shutdownWG.Done()
		processor.modelUpdateLoop()
	}()
```

Bit of a style preference, but I find that keeping the Add and Done calls close together makes it easier to reason about.

axw reviewed (approved):
Looks good, just a couple of minor suggestions.

VenuEmmadi committed:
fix(processor/isolationforest): address code review feedback

- Use anonymous function pattern for WaitGroup management in goroutines
- Apply context.WithoutCancel() for test cleanup to prevent race conditions
- Move WaitGroup.Done() to anonymous function wrapper for panic safety
- Remove defer from modelUpdateLoop() method for cleaner separation

Addresses reviewer feedback on goroutine lifecycle management and test context handling.

andrzej-stencel reviewed (approved):
Thanks @VenuEmmadi for this fix.

As a side note, it seems that [starting the model update loop](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/94af87394e863fa9c1b243e70ae3b04ef2e6cc51/processor/isolationforestprocessor/processor.go#L105-L106) should happen in the processor's `Start` function rather than in the `newIsolationForestProcessor` function.

VenuEmmadi committed:
fix(processor/isolationforest): start background goroutine in Start method

Move model update loop initialization from newIsolationForestProcessor
to Start() method following OpenTelemetry component lifecycle patterns.

VenuEmmadi committed:
Merge branch 'main' into fix/isolation-forest-goroutine-leak

VenuEmmadi committed:
fix(processor/isolationforest): start background goroutine in Start method

Move model update loop initialization from newIsolationForestProcessor
to Start() method following OpenTelemetry component lifecycle patterns.
